### PR TITLE
Unrolled Union instances

### PR DESCRIPTION
**Pull request by robrix**
This PR:

- Redefines the `Union` instances by unrolling the clauses.
- ~~🔥s the redundant injection parameters to the `'` versions of the methods.~~
- ~~Renames the methods to drop the `'`.~~
- Defines `Apply1` instances for `Union`s of up to 150 functors.
- 🔥s use of `PolyKinds`.
- 🔥s `Apply0`.
- Defines `apply2` as a function (not a method), in terms of `apply`.
- Greatly simplifies the TH generating the `Apply` instances.
- Greatly speeds up compilation of `Data.Union`.

### CODE REVIEW

File: src/Data/Union.hs [hunk 1]
```diff
@@ -139,42 +134,20 @@ instance {-# OVERLAPPING #-} t :< r => t :< (t' ': r) where
...
+  apply1 :: proxy c -> (forall g . (c g, g :< fs) => g a -> b) -> Union fs a -> b
 
-instance (c f0) => Apply1 c '[f0] where
-  apply1' _ f (Union _ r) = f (Union 0) (unsafeCoerce r :: f0 a)
+  apply1_2 :: proxy c -> (forall g . (c g, g :< fs) => g a -> g b -> d) -> Union fs a -> Union fs b -> Maybe d
 
-  apply1_2' _ f (Union _ r1) (Union _ r2) = Just (f (Union 0) (unsafeCoerce r1 :: f0 a) (unsafeCoerce r2))
 
-mkApply1Instances [2..55]
+mkApply1Instances [1..150]
```
Comment (robrix) [inline src/Data/Union.hs]: This subsumes @joshvera’s `increase-instances` branch.

File: src/Data/Union.hs [hunk 2]
```diff
@@ -139,42 +134,20 @@ instance {-# OVERLAPPING #-} t :< r => t :< (t' ': r) where
...
-apply1 proxy f = apply1' proxy (const f)
-
-apply1_2 :: Apply1 c fs => proxy c -> (forall g . c g => g a -> g b -> d) -> Union fs a -> Union fs b -> Maybe d
-apply1_2 proxy f = apply1_2' proxy (const f)
-
+  apply1 :: proxy c -> (forall g . (c g, g :< fs) => g a -> b) -> Union fs a -> b
 
-instance (c f0) => Apply1 c '[f0] where
-  apply1' _ f (Union _ r) = f (Union 0) (unsafeCoerce r :: f0 a)
+  apply1_2 :: proxy c -> (forall g . (c g, g :< fs) => g a -> g b -> d) -> Union fs a -> Union fs b -> Maybe d
```
Comment (robrix) [inline src/Data/Union.hs]: We no longer need the reinjection parameter ✨

File: stack.yaml [hunk 3]
```diff
@@ -3,4 +3,4 @@ extra-package-dbs: []
 packages:
 - '.'
 extra-deps: []
-resolver: lts-6.10
+resolver: nightly-2017-09-10
```
Comment (robrix) [inline stack.yaml]: ghc-8.2.1 ✨

File: src/Data/Union.hs [hunk 4]
```diff
@@ -139,42 +134,20 @@ instance {-# OVERLAPPING #-} t :< r => t :< (t' ': r) where
...
-
-  apply0_2' :: proxy c -> (forall g . c (g a) => (forall x. g x -> Union fs x) -> g a -> g b -> d) -> Union fs a -> Union fs b -> Maybe d
-
-apply0 :: Apply0 c fs a => proxy c -> (forall g . c (g a) => g a -> b) -> Union fs a -> b
-apply0 proxy f = apply0' proxy (const f)
-
-apply0_2 :: Apply0 c fs a => proxy c -> (forall g . c (g a) => g a -> g b -> d) -> Union fs a -> Union fs b -> Maybe d
-apply0_2 proxy f = apply0_2' proxy (const f)
-
-instance (c (f0 a)) => Apply0 c '[f0] a where
```
Comment (robrix) [inline src/Data/Union.hs]: We now define the first instance using the same templates as the rest.

**Comment by robrix** [comment]
This is untenable; attempting to compile a profiling build with only 75 instances brought my machine to its knees and I eventually had to kill it and reboot the machine.

**Comment by robrix** [comment]
I’ve now made a few more changes that have made this tractable:

1. I’ve turned off `PolyKinds`. This saved 8s with instances for up to 50 types all by itself.

2. The templates are vastly simpler, which also means quicker to instantiate.

3. `Apply` now has one and only one method, `apply`. `apply2` is defined in terms of `apply`, as are the `apply'` and `apply2'` variants (which are back again; see below for details).

4. The `apply` and `apply2` functions no longer provide a `g :< fs` constraint in the context of the lifted function, necessitating the return of `apply'` and `apply2'`.

I spent quite a bit of time trying to find a constant-time way to select the _n_th element of a type-level list, and I’m now fairly well convinced that it can’t be done. Fortunately, however, this didn’t end up being the problem at all; coercions did.

Generating `Apply` instances for up to 70 types with `g :< fs` in the context yielded these stats (from `-ddump-simpl`) after about 50s:

```
Result size of Tidy Core
  = {terms: 19,992, types: 99,567, coercions: 1,007,578, joins: 0/22}
```

Without it, after about 16s:

```
Result size of Tidy Core
  = {terms: 7,917, types: 97,152, coercions: 47,213, joins: 0/22}
```

I believe we were running afoul of a known `ghc` issue described as [“coercion pile-up”](https://ghc.haskell.org/trac/ghc/wiki/Performance/Compiler#Coercionpile-up). With all of the fixes in this PR, we can now compile instances for up to 150 types in 78s on my machine, with these stats:

```
Result size of Tidy Core
  = {terms: 72,111, types: 343,396, coercions: 183,160, joins: 0/22}
```

and of course when you don’t produce stats, it’s even quicker; 52s.

**Comment by joshvera** [comment]
👍

**Comment by joshvera** [comment]
👍

**Pull request merged**